lib/utils/replace.go

```diff
@@ -221,6 +221,109 @@ func KubeResourceMatchesRegex(input types.KubernetesResource, resources []types.
 	return false, nil
 }
 
+// KubeResourceCouldMatchRules assess whether the user is permitted to perform its request
+// based on the defined kubernetes_resource rules. The aim is to catch cases when the user
+// has no access and present then a more user-friendly error message instead of returning
+// an empty list.
+// This function is not responsible for enforcing access rules.
+func KubeResourceCouldMatchRules(input types.KubernetesResource, resources []types.KubernetesResource, cond types.RoleConditionType) (bool, error) {
+	if len(input.Verbs) != 1 {
+		return false, trace.BadParameter("only one verb is supported, input: %v", input.Verbs)
+	}
+	if input.Name != "" {
+		return false, trace.BadParameter("name is not supported for KubeResourceMatchesNamespaceRegex")
+	}
+
+	verb := input.Verbs[0]
+	isDeny := cond == types.Deny
+
+	// isClusterWideResource is true if the resource is cluster-wide, e.g. a
+	// namespace resource or a clusterrole.
+	isClusterWideResource := slices.Contains(types.KubernetesClusterWideResourceKinds, input.Kind)
+
+	// If the user is allowed to list/read/watch a resource, they should be able to see the
+	// namespace in which the resource is.
+	// This is a special case because we don't want to require the user to have
+	// access to the namespace resource itself.
+	// This is only allowed for the list/read/watch verbs because we don't want
+	// to allow the user to create/update/delete a namespace they don't have
+	// permissions for.
+	targetsReadOnlyNamespace := input.Kind == types.KindKubeNamespace &&
+		slices.Contains([]string{types.KubeVerbGet, types.KubeVerbList, types.KubeVerbWatch}, verb)
+
+	for _, resource := range resources {
+		// If the resource has a wildcard verb, it matches all verbs.
+		// Otherwise, the resource must have the verb we're looking for otherwise
+		// it doesn't match.
+		// When the resource has a wildcard verb, we only allow one verb in the
+		// resource input.
+		if !isVerbAllowed(resource.Verbs, verb) {
+			continue
+		}
+		switch {
+		// If the user has access to a specific namespace, they should be able to
+		// access all resources in that namespace.
+		case resource.Kind == types.KindKubeNamespace:
+			cond := !isDeny || resource.Name == types.Wildcard
```

Comment:
We already have `cond` as a parameter of this function, maybe some other name just to make sure it's not confused with the first one?